internal/user_validator.go

```diff
@@ -90,24 +95,62 @@ func (i *ValidateUser) Setup(ctx context.Context) error {
 
 func decodePgpKey(pgpKey string) (*openpgp.Entity, error) {
 	pgpKey = strings.TrimRight(pgpKey, " \n\r")
-	if strings.Contains(pgpKey, " ") {
-		return nil, fmt.Errorf("PGP key has spaces in it")
+	pgpKey = strings.TrimSpace(pgpKey)
+
+	var errArmoredKey = errors.New("ASCII-armored PGP keys are not supported; please remove type headers and checksum")
+
+	keyArmor := func(anchor string) string {
+		return fmt.Sprintf("-----%s %s-----", anchor, openpgp.PublicKeyType)
 	}
+	keyArmorStart := keyArmor("BEGIN")
 
-	suffix := ""
-	for i := 0; i < strings.Count(pgpKey, "="); i++ {
-		suffix += "="
+	if strings.HasPrefix(pgpKey, keyArmorStart[:strings.Index(keyArmorStart, " ")]) {
+		return nil, errArmoredKey
 	}
-	if !strings.HasSuffix(pgpKey, suffix) {
-		return nil, fmt.Errorf("Equals signs are not add the end")
+
+	if strings.Contains(pgpKey, " ") {
+		return nil, fmt.Errorf("given PGP key cannot contain spaces")
 	}
 
 	data, err := base64.StdEncoding.DecodeString(pgpKey)
 	if err != nil {
-		return nil, fmt.Errorf("error decoding given PGP key: %w", err)
+		// Save the original Base64 decoder error,
+		// to return if an error is not related to
+		// ASCII armor parsing or validation.
+		decodeErr := err
+
+		pgpKey = fmt.Sprintf("%s\n\n%s\n%s", keyArmorStart, pgpKey, keyArmor("END"))
+		block, err := armor.Decode(strings.NewReader(pgpKey))
+		if err != nil {
+			return nil, fmt.Errorf("error decoding given ASCII-armored PGP key: %w", err)
+		}
+
+		// Drain the Reader buffer, which causes the CRC24
+		// checksum to be computed for the given ASCII armor.
+		_, err = io.Copy(ioutil.Discard, block.Body)
+		if err != nil {
+			if _, ok := err.(pgperr.StructuralError); ok {
+				return nil, fmt.Errorf("error decoding given ASCII-armored PGP key: %w", err)
+			}
+			return nil, fmt.Errorf("error decoding given PGP key: %w", decodeErr)
+		}
+		Log().Debugln("A valid ASCII-armored PGP key has been given")
+
+		return nil, errArmoredKey
```

Comment:
This is the outcome we are doing this entire change for: users providing a key with a checksum and us telling them to remove it.

Reply:
@geoberle, the error message has been updated to notify the user about how the encoded CRC24 checksum looks like, so it should be easier to find it, should someone would copy it from an ASCII-armored key block (i.e., `gpg --export --armor` output).